nano/secure/ledger.cpp

```diff
@@ -826,26 +826,47 @@ nano::uint128_t nano::ledger::balance_safe (nano::transaction const & transactio
 }
 
 // Balance for an account by account number
-nano::uint128_t nano::ledger::account_balance (nano::transaction const & transaction_a, nano::account const & account_a)
+nano::uint128_t nano::ledger::account_balance (nano::transaction const & transaction_a, nano::account const & account_a, bool only_confirmed_a)
 {
 	nano::uint128_t result (0);
-	nano::account_info info;
-	auto none (store.account_get (transaction_a, account_a, info));
-	if (!none)
+	if (only_confirmed_a)
+	{
+		nano::confirmation_height_info info;
+		if (!store.confirmation_height_get (transaction_a, account_a, info))
+		{
+			result = balance (transaction_a, info.frontier);
+		}
+	}
+	else
 	{
-		result = info.balance.number ();
+		nano::account_info info;
+		auto none (store.account_get (transaction_a, account_a, info));
+		if (!none)
+		{
+			result = info.balance.number ();
+		}
 	}
 	return result;
 }
 
-nano::uint128_t nano::ledger::account_pending (nano::transaction const & transaction_a, nano::account const & account_a)
+nano::uint128_t nano::ledger::account_pending (nano::transaction const & transaction_a, nano::account const & account_a, bool only_confirmed_a)
 {
 	nano::uint128_t result (0);
 	nano::account end (account_a.number () + 1);
 	for (auto i (store.pending_begin (transaction_a, nano::pending_key (account_a, 0))), n (store.pending_begin (transaction_a, nano::pending_key (end, 0))); i != n; ++i)
 	{
 		nano::pending_info const & info (i->second);
-		result += info.amount.number ();
+		if (only_confirmed_a)
+		{
+			if (block_confirmed (transaction_a, i->first.hash))
```

Comment:
I implemented something similar for a PR I'm working on, though I added a pruning check - that's need right? Something like

`if (block_confirmed (transaction_a, i->first.hash) || (pruning && store.pruned_exists (transaction_a, i->first.hash)))`

Reply:
Oh really! Good to see you saw value in it too 👍 Oh yeh probably, wonder if it should be part of block_confirmed. There's also the conf height processor to consider too (`is_being_confirmed`). Will have a look, thanks!

Reply:
So I think just checking no conf height in progress checking is fine. We only post callbacks after the block is confirmed so could cause some confusion otherwise. Added the pruning check though thanks! We also shouldn't error in account_info if not using include_only_confirmed, as there are still valid non-confirmed entries. The docs should also make clear that "include_only_confirmed" only affects balance and pending count (if requested).